File: MiniDOMOffice/CRM/Chat/Rooms/Rooms.vb
Imports minidom
Imports minidom.Sistema
Imports minidom.Databases
Imports minidom.Internals
Imports minidom.Messenger

Namespace Internals
    ''' <summary>
    ''' Stanze
    ''' </summary>
    ''' <remarks></remarks>
    Public NotInheritable Class CChatRoomsClass
        Inherits CModulesClass(Of CChatRoom)

        Private m_MainRoom As CChatRoom

        Public Sub New()
            MyBase.New("modChatRooms", GetType(CChatRoomCursor), -1)
        End Sub

        Public Function GetItemByName(ByVal name As String) As CChatRoom
            name = Strings.Trim(name)
            For Each room As CChatRoom In Me.LoadAll
                If (room.Stato = ObjectStatus.OBJECT_VALID AndAlso Strings.Compare(room.Name, name, CompareMethod.Text) = 0) Then Return room
            Next
            Return Nothing
        End Function

        Public ReadOnly Property MainRoom As CChatRoom
            Get
                If (Me.m_MainRoom Is Nothing) Then
                    Me.m_MainRoom = Me.GetItemByName("main")
                End If
                If (Me.m_MainRoom Is Nothing) Then
                    Me.m_MainRoom = New CChatRoom()
                    Me.m_MainRoom.Name = "main"
                    Me.m_MainRoom.Stato = ObjectStatus.OBJECT_VALID
                    Me.m_MainRoom.Save()
                    For Each u As CUser In Sistema.Users.LoadAll
                        If (u.Stato = ObjectStatus.OBJECT_VALID) Then
                            Me.m_MainRoom.AddMember(u)
                        End If
                    Next
                End If
                Return Me.m_MainRoom
            End Get
        End Property

    End Class
End Namespace

Partial Public Class Messenger



    Private Shared m_Rooms As CChatRoomsClass = Nothing

    ''' <summary>
    ''' Accede alle stanze
    ''' </summary>
    ''' <value></value>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public Shared ReadOnly Property Rooms As CChatRoomsClass
        Get
            If (m_Rooms Is Nothing) Then m_Rooms = New CChatRoomsClass
            Return m_Rooms
        End Get
    End Property

End Class
 

File: MiniDOMOffice/CRM/Chat/Rooms/Rooms.cs
using DMD;
using DMD.Databases;
using static minidom.Messenger;
using minidom.repositories;

namespace minidom
{
    namespace repositories
    {

        /// <summary>
        /// Repository di oggetti di tipo <see cref="CChatRoom"/>
        /// </summary>
        /// <remarks></remarks>
        public sealed class CChatRoomsClass
            : CModulesClass<Messenger.CChatRoom>
        {
            private Messenger.CChatRoom m_MainRoom;


            /// <summary>
            /// Costruttore
            /// </summary>
            public CChatRoomsClass() 
                : base("modChatRooms", typeof(Messenger.CChatRoomCursor), -1)
            {
            }

            /// <summary>
            /// Restituisce la stanza con nome
            /// </summary>
            /// <param name="name"></param>
            /// <returns></returns>
            public Messenger.CChatRoom GetItemByName(string name)
            {
                name = DMD.Strings.Trim(name);
                foreach (Messenger.CChatRoom room in LoadAll())
                {
                    if (room.Stato == ObjectStatus.OBJECT_VALID && DMD.Strings.Compare(room.Name, name, true) == 0)
                        return room;
                }

                return null;
            }

            /// <summary>
            /// Restituisce un riferimento alla stanza principale
            /// </summary>
            public Messenger.CChatRoom MainRoom
            {
                get
                {
                    if (m_MainRoom is null)
                    {
                        m_MainRoom = GetItemByName("main");
                    }

                    if (m_MainRoom is null)
                    {
                        m_MainRoom = new Messenger.CChatRoom();
                        m_MainRoom.Name = "main";
                        m_MainRoom.Stato = ObjectStatus.OBJECT_VALID;
                        m_MainRoom.Save();
                        foreach (Sistema.CUser u in Sistema.Users.LoadAll())
                        {
                            if (u.Stato == ObjectStatus.OBJECT_VALID)
                            {
                                m_MainRoom.AddMember(u);
                            }
                        }
                    }

                    return m_MainRoom;
                }
            }
        }
    }

    public partial class Messenger
    {

        private static CChatRoomsClass m_Rooms = null;

        /// <summary>
        /// Repository di oggetti di tipo <see cref="CChatRoom"/>
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
        public static CChatRoomsClass Rooms
        {
            get
            {
                if (m_Rooms is null)
                    m_Rooms = new CChatRoomsClass();
                return m_Rooms;
            }
        }
    }
}